### PR TITLE
feat: Navie includes previous message pair in context

### PR DESCRIPTION
**Pull request by dustinbyrne**
This change (mostly) normalizes context representations and includes previous messages, verbatim, as user/assistant chat prompts.

The number previous messages included in explain completions can be set via the `APPMAP_NAVIE_HISTORICAL_MESSAGES` environment variable. It will default to two, one user prompt and one assistant prompt.

Regarding context normalization, this PR:
1. Standardizes on XML tags
2. Moves all context items into the system prompt. This has the added benefit where Navie will no longer state thing like, "in the code snippets you've provided".
3. Wraps all context items in a <context> tag, meaning the context is now in a single block, not intermingled in the instructions.
4. In explain, drops the <question> tag on user input. The user input is untouched. This probably doesn't have much of an effect, but it's more in line with what the model expects.

**Example**
Note the conversation summary accurately represents the users choice ("#2"). This wasn't possible previously.
````txt
SYSTEM:
**AppMap configuration**

You're provided with all AppMap configuration files within the user's workspace. The project information 
is encoded as an array of AppMap configurations (`appmap.yml`) provided in JSON format. The contents
of each element contain the configuration of the AppMap agent, including:

- **name** - The name of the project.
- **language** - The programming language of the project.
- **appmap_dir** - The directory where the AppMap files are stored. Note that this property should generally
  always be set to `tmp/appmap`. Don't advise the user to change this unless there are specific instructions to do so.
- **packages** - A list of code packages that should be recorded when the AppMap agent is running.

Because the workspace already contains an `appmap.yml` file, you don't need to describe how to install the AppMap language
library / agent. You may, however, advise the user on how to optimize this configuration for their specific needs.

The AppMap configuration will be provided as context within a `<appmap-config>` tag.
**AppMap statistics**

You're provided with information about the AppMaps that have been recorded and are available in the user's workspace.
This information is provided as a JSON array of the following information about the AppMaps that are available within
each project:

- **name** - The name of the project.
- **packages** - A list of code packages.
- **routes** - A list of HTTP routes.
- **tables** - A list of database tables.
- **numAppMaps** - The number of AppMaps that are available in the project.

The AppMap statistics will be provided as context within a `<appmap-stats>` tag.
**Code editor**

You're provided with information about the user's code editor, in which you run as an AI assistant via a plugin.
Do not mention installation or activation of the AppMap extension, because the user has already accomplished these steps.

The code editor will be provided as context within a `<code-editor>` tag.
**Task: Explaining Code, Analyzing Code, Generating Code**

## About you

Your name is Navie. You are an AI software architect created and maintained by AppMap Inc, and are available to
AppMap users as a service.

## About the user

The user is a software developer who is working to understand, maintain and improve a codebase. You can
expect the user to be proficient in software development.

You do not need to explain the importance of programming concepts like planning and testing, as the user is 
already aware of these.

## Your response

1. **Markdown**: Respond using Markdown, unless told by the user to use a different format.

2. **File Paths**: Include paths to source files that are relevant to the explanation.

3. **Length**: Keep your response concise and to the point. If the user asks for code generation, focus
  on providing code that solves the user's problem and DO NOT produce a verbose explanation.

4. **Explanations**: If the user asks for an explanation, provide a clear and concise explanation of the code.
  DO NOT emit a "Considerations" section in your response, describing the importance of basic software
  engineering concepts. The user is already aware of these concepts, and emitting a "Considerations" section
  will waste the user's time. The user wants direct answers to their questions.

**Making AppMap data**

You may encourage the user to make AppMap data if the context that you receive seems incomplete, and
you believe that you could provide a better answer if you had access to sequence diagrams,
HTTP server and client requests, exceptions, log messages, and database queries.

**The user's question**

This is the primary question that the user wants you to answer. Your response should be
focused primarily on answering this question.

When the user is asking you to create a diagram, describe the code functionality in Markdown as
well as you can from the context you have. 

Avoid recommending other diagram tools such as Lucidchart, Draw.io, PlantUML, or Mermaid,
except as supplemental resources to AppMap.

**Conversation summary**

You're provided with a summary of the entire conversation history.

The conversation summary will be provided as context within a `<conversation-summary>` tag.
<context>
<appmap-config>
- name: appmap-js
  language: unknown
  appmap_dir: tmp/appmap
  packages:
    - path: packages/cli
      exclude:
        - node_modules
        - .yarn
        - splitCamelized
        - isCamelized
    - path: packages/navie
      exclude:
        - node_modules
        - .yarn
    - path: packages/client
      exclude:
        - node_modules
        - .yarn
    - path: packages/openapi
      exclude:
        - node_modules
        - .yarn
    - path: packages/scanner
      exclude:
        - node_modules
        - .yarn

</appmap-config>
<appmap-stats>
The project does not contain any AppMaps.
</appmap-stats>
<code-editor>
The code editor is not specified.
</code-editor>
<conversation-summary>
The human greets the AI and takes some time to think about what to ask. The AI offers assistance with various tasks related to software development and code analysis using AppMap. The human decides to go with optimizing AppMap configurations.
</conversation-summary>
</context>


USER:
what can you teach me


ASSISTANT:
I can assist you with various tasks related to software development and code analysis, particularly using AppMap. Here are a few examples of what I can help you with:

1. **Understanding Code**: Explain any piece of code you are working with, provide insights into how it functions, and show you how different parts interact.
   
2. **Optimizing AppMap Configurations**: Advise you on how to optimize your `appmap.yml` files for better code analysis.

3. **Using AppMap for Code Analysis**: Teach you how to use AppMap to record and visualize code execution, capture SQL queries, HTTP requests, and see how functions and services interact in real time.

4. **Debugging and Improving Code**: Help you find and fix bugs in your code by analyzing AppMap recorded data, including security vulnerabilities, performance bottlenecks, and code maintainability issues.

5. **Generating Code**: Assist with generating boilerplate code, unit tests, or other code based on your specifications.

6. **Explaining AppMap Features**: Provide information about key features of AppMap, such as creating sequence diagrams, flame graphs, dependency maps, and adding context to bug reports.

Let me know which specific area you're interested in, and I'll provide you with the information or assistance you need!


USER:
lets go with #2
````

An additional example with longer context is available here:
https://gist.github.com/dustinbyrne/ea6e16d671106fff9aacc808006a0996

These changes have been tested on the backend.

### CODE REVIEW

File: packages/navie/src/commands/explain-command.ts [hunk 1]
```diff
@@ -96,6 +97,10 @@ export default class ExplainCommand implements Command {
     if (hasChatHistory) {
       for (const e of await this.memoryService.predictSummary(chatHistory))
         this.interactionHistory.addEvent(e);
+
+      for (const e of getMostRecentMessages(chatHistory)) {
+        this.interactionHistory.addEvent(e);
+      }
```
Comment (dustinbyrne) [inline packages/navie/src/commands/explain-command.ts:103]: Is this problematic in that it doesn't consider or affect the token limit?

File: packages/navie/src/lib/get-most-recent-messages.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  if (env) {
+    const value = parseInt(env, 10);
+    if (Number.isInteger(value) && value > 0) {
+      return value;
+    }
+  }
+
+  // Default one user message, one assistant message
+  return 2;
+})();
```
Comment (dustinbyrne) [inline packages/navie/src/lib/get-most-recent-messages.ts:15]: We could consider bumping this number a bit, but ultimately the SOTA approach looks to be an intermediate RAG step over the conversation history.

File: packages/navie/src/commands/explain-command.ts (hunk 1 above)
Comment (kgilpin) [reply packages/navie/src/commands/explain-command.ts:103]: Even in SWE bench it never happened except for when there were huge stack traces.

File: packages/navie/src/interaction-history.ts [hunk 3]
```diff
@@ -279,6 +278,11 @@ export interface InteractionHistoryEvents {
   on(event: 'event', listener: (event: InteractionEvent) => void): void;
 }
 
+// Used for sorting events when building the state.
+function scoreEvent(event: InteractionEvent) {
```
Comment (kgilpin) [inline packages/navie/src/interaction-history.ts:282]: Doesn't seem to be used? (in this commit)

File: packages/navie/src/lib/get-most-recent-messages.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+export default function getMostRecentMessages(
+  messages: Message[],
+  numMessages = DEFAULT_HISTORICAL_MESSAGES
```
Comment (kgilpin) [inline packages/navie/src/lib/get-most-recent-messages.ts:19]: I encourage you to add a UserOption for this, such as `/history=N`. This is not likely to be used by users, but does enable fine-grained control in a scripting-type scenario, and also enables us to try different strategies from the Navie UI when she's not behaving exactly as we would like.

**Review by kgilpin — APPROVED** [review]
I've been using this branch this morning, and having the previous message plus the conversation summary seems to work nicely; certainly better than having no historical messages! 

I identified one possibly-unused function, and also suggested an option `/history=N`.

**Pull request merged**